x/pki/handler_test.go

```diff
@@ -404,107 +489,704 @@ func TestHandler_AddX509Cert_ForFailedCertificateVerification(t *testing.T) {
+	setup := Setup()
+
+	// add root x509 certificate
+	proposeAndApproveRootCertificate(t, &setup, setup.Trustee)
+
+	// add intermediate x509 certificate
+	addIntermediateX509Cert := types.NewMsgAddX509Cert(constants.IntermediateCertPem, setup.Trustee)
+	result := setup.Handler(setup.Ctx, addIntermediateX509Cert)
+	require.Equal(t, sdk.CodeOK, result.Code)
+
+	// add leaf x509 certificate
+	addLeafX509Cert := types.NewMsgAddX509Cert(constants.LeafCertPem, setup.Trustee)
+	result = setup.Handler(setup.Ctx, addLeafX509Cert)
+	require.Equal(t, sdk.CodeOK, result.Code)
+
+	// query root certificate
+	rootCertificate, _ := querySingleCertificate(&setup, constants.RootSubject, constants.RootSubjectKeyID)
+	require.Equal(t, constants.RootCertPem, rootCertificate.PemCert)
+
+	// check child certificate identifiers of root certificate
+	rootCertChildren := setup.PkiKeeper.GetChildCertificates(setup.Ctx,
+		constants.RootSubject, constants.RootSubjectKeyID)
+
+	require.Equal(t, 1, len(rootCertChildren.CertIdentifiers))
+	require.Equal(t,
+		types.NewCertificateIdentifier(constants.IntermediateSubject, constants.IntermediateSubjectKeyID),
+		rootCertChildren.CertIdentifiers[0])
+
+	// query intermediate certificate
+	intermediateCertificate, _ :=
+		querySingleCertificate(&setup, constants.IntermediateSubject, constants.IntermediateSubjectKeyID)
+	require.Equal(t, constants.IntermediateCertPem, intermediateCertificate.PemCert)
+
+	// check child certificate identifiers of intermediate certificate
+	intermediateCertChildren := setup.PkiKeeper.GetChildCertificates(setup.Ctx,
+		constants.IntermediateSubject, constants.IntermediateSubjectKeyID)
+
+	require.Equal(t, 1, len(intermediateCertChildren.CertIdentifiers))
+	require.Equal(t,
+		types.NewCertificateIdentifier(constants.LeafSubject, constants.LeafSubjectKeyID),
+		intermediateCertChildren.CertIdentifiers[0])
+
+	// query leaf certificate
+	leafCertificate, _ := querySingleCertificate(&setup, constants.LeafSubject, constants.LeafSubjectKeyID)
+	require.Equal(t, constants.LeafCertPem, leafCertificate.PemCert)
+
+	// check child certificate identifiers of leaf certificate
+	leafCertChildren :=
+		setup.PkiKeeper.GetChildCertificates(setup.Ctx, constants.LeafSubject, constants.LeafSubjectKeyID)
+
+	require.Equal(t, 0, len(leafCertChildren.CertIdentifiers))
+}
+
+//nolint:funlen
+func TestHandler_AddX509Cert_EachChildCertRefersToTwoParentCerts(t *testing.T) {
 	setup := Setup()
 
 	// store root certificate
 	rootCert := rootCertificate(setup.Trustee)
+
 	setup.PkiKeeper.AddApprovedCertificate(setup.Ctx, rootCert)
+	setup.PkiKeeper.SetUniqueCertificateKey(setup.Ctx, rootCert.Subject, rootCert.SerialNumber)
 
 	// store second root certificate
 	rootCert = rootCertificate(setup.Trustee)
 	rootCert.SerialNumber = SerialNumber
+
 	setup.PkiKeeper.AddApprovedCertificate(setup.Ctx, rootCert)
+	setup.PkiKeeper.SetUniqueCertificateKey(setup.Ctx, rootCert.Subject, rootCert.SerialNumber)
```

Comment:
Shouldn't it be an internal logic of the keeper?

Reply:
Created a separate ticket for this: https://github.com/zigbee-alliance/distributed-compliance-ledger/issues/51